pkg/metadata/container.go

```diff
@@ -0,0 +1,183 @@
+	// ID is the container id.
+	ID string
+	// Name is the container name.
+	Name string
+	// SandboxID is the sandbox id the container belongs to.
+	SandboxID string
+	// Config is the CRI container config.
+	Config *runtime.ContainerConfig
+	// ImageRef is the reference of image used by the container.
+	ImageRef string
+	// Pid is the init process id of the container.
+	Pid uint32
+	// CreatedAt is the created timestamp.
+	CreatedAt int64
+	// StartedAt is the started timestamp.
+	StartedAt int64
+	// FinishedAt is the finished timestamp.
+	FinishedAt int64
+	// ExitCode is the container exit code.
+	ExitCode int32
+	// CamelCase string explaining why container is in its current state.
+	Reason string
+	// Human-readable message indicating details about why container is in its
+	// current state.
+	Message string
+	// Removing indicates that the container is in removing state.
+	// In fact, this field doesn't need to be checkpointed.
+	// TODO(random-liu): Skip this during serialization when we put object
+	// into the store directly.
+	// TODO(random-liu): Reset this field to false during state recoverry.
+	Removing bool
+}
+
+// State returns current state of the container based on the metadata.
+func (c *ContainerMetadata) State() runtime.ContainerState {
+	if c.FinishedAt != 0 {
+		return runtime.ContainerState_CONTAINER_EXITED
+	}
+	if c.StartedAt != 0 {
+		return runtime.ContainerState_CONTAINER_RUNNING
+	}
+	if c.CreatedAt != 0 {
+		return runtime.ContainerState_CONTAINER_CREATED
+	}
+	return runtime.ContainerState_CONTAINER_UNKNOWN
+}
+
+// ContainerUpdateFunc is the function used to update ContainerMetadata.
+type ContainerUpdateFunc func(ContainerMetadata) (ContainerMetadata, error)
+
+// ContainerToStoreUpdateFunc generates a metadata store UpdateFunc from ContainerUpdateFunc.
+func ContainerToStoreUpdateFunc(u ContainerUpdateFunc) store.UpdateFunc {
+	return func(data []byte) ([]byte, error) {
+		meta := &ContainerMetadata{}
+		if err := json.Unmarshal(data, meta); err != nil {
+			return nil, err
+		}
+		newMeta, err := u(*meta)
+		if err != nil {
+			return nil, err
+		}
+		return json.Marshal(newMeta)
+	}
+}
+
+// ContainerStore is the store for metadata of all containers.
+type ContainerStore interface {
+	// Create creates a container from ContainerMetadata in the store.
+	Create(ContainerMetadata) error
+	// Get gets a specified container.
+	Get(string) (*ContainerMetadata, error)
+	// Update updates a specified container.
+	Update(string, ContainerUpdateFunc) error
+	// List lists all containers.
+	List() ([]*ContainerMetadata, error)
+	// Delete deletes the container from the store.
+	Delete(string) error
+}
+
+// containerStore is an implmentation of ContainerStore.
+type containerStore struct {
+	store store.MetadataStore
+}
+
+// NewContainerStore creates a ContainerStore from a basic MetadataStore.
+func NewContainerStore(store store.MetadataStore) ContainerStore {
+	return &containerStore{store: store}
+}
+
```

Comment:
I am going to skip the rest of the file assuming that they are just the sandbox metadata store with name replaced.